### PR TITLE
boot: add bootloader options to coreKernel

### PR DESCRIPTION
**Pull request by mvo5**
This allows us to ask the device if we are on UC20 essentially, and figure out
what options we should use when searching for a bootloader to extract/remove
kernel assets from.

This is a small subset of PR#8001 that is useful on it's own.

### CODE REVIEW

File: boot/kernel_os.go [hunk 1]
```diff
@@ -53,7 +53,8 @@ func (bp *coreBootParticipant) SetNextBoot() (rebootRequired bool, err error) {
 }
 
 type coreKernel struct {
-	s snap.PlaceInfo
+	s   snap.PlaceInfo
+	dev Device
```
Comment (pedronis) [inline boot/kernel_os.go]: I don't think we should carry a dev at this level, we should probably carry already a bootloader.Options or something like that

File: boot/kernel_os.go [hunk 2]
```diff
@@ -73,11 +79,15 @@ func (k *coreKernel) RemoveKernelAssets() error {
 }
 
 func (k *coreKernel) ExtractKernelAssets(snapf snap.Container) error {
-	bootloader, err := bootloader.Find("", nil)
+	opts := &bootloader.Options{}
+	if k.dev.HasModeenv() {
+		// we are in uc20 mode, so find an extractable kernel
+		opts.ExtractedRunKernelImage = true
```
Comment (pedronis) [inline boot/kernel_os.go]: afaict this bit is not tested

File: boot/kernel_os.go [hunk 3]
```diff
@@ -63,7 +64,12 @@ func (*coreKernel) IsTrivial() bool { return false }
 
 func (k *coreKernel) RemoveKernelAssets() error {
 	// XXX: shouldn't we check the snap type?
-	bootloader, err := bootloader.Find("", nil)
+	opts := &bootloader.Options{}
+	if k.dev.HasModeenv() {
+		// we are in uc20 mode, so find an extractable kernel
+		opts.ExtractedRunKernelImage = true
```
Comment (pedronis) [inline boot/kernel_os.go]: afaict this bit is not tested here, nor in the larger PR

File: boot/kernel_os.go (hunk 1 above)
Comment (anonymouse64) [reply boot/kernel_os.go]: hmm I guess the reason that's weird to me is that where we create the coreKernel we don't have state about the bootloader and that type of thing in the overlord managers, but what we do have that's readily accessible is the model, which implements boot.Device. What do you think we should use instead?

File: boot/kernel_os.go (hunk 3 above)
Comment (anonymouse64) [reply boot/kernel_os.go]: hmm I suppose this method isn't called in the CoreKernel test here: https://github.com/snapcore/snapd/pull/8001/files#diff-bf7a36650e7c3d6180b9547be1e9712eR374-R396

I will add something tomorrow if @mvo5 doesn't beat me to it :-)

File: boot/kernel_os.go (hunk 1 above)
Comment (bboozzoo) [reply boot/kernel_os.go]: I've pushed an update that adds the whole `bootloader.Options` to `coreKernel`. Could have limited it to just the extract-run-kernel flag, but it makes for fewer switches in the actual `*KernelAssets()` methods.

File: boot/kernel_os.go (hunk 3 above)
Comment (bboozzoo) [reply boot/kernel_os.go]: I've cherry-picked, edited and pushed fafd3b526fc15061a4900c62b5e9abffcfd0078b.

File: boot/boot.go [hunk 4]
```diff
@@ -96,12 +97,21 @@ func Participant(s snap.PlaceInfo, t snap.Type, dev Device) BootParticipant {
 	return trivial{}
 }
 
+// bootloaderOptionsForDeviceKernel returns a set of bootloader options that
+// enable correct kernel extraction and removal for given device
+func bootloaderOptionsForDeviceKernel(dev Device) bootloader.Options {
```
Comment (pedronis) [inline boot/boot.go]: I would return a *

Comment (pedronis) [inline boot/boot.go]: I would return a *

File: boot/kernel_os.go [hunk 5]
```diff
@@ -53,7 +53,8 @@ func (bp *coreBootParticipant) SetNextBoot() (rebootRequired bool, err error) {
 }
 
 type coreKernel struct {
-	s snap.PlaceInfo
+	s     snap.PlaceInfo
+	bopts bootloader.Options
```
Comment (pedronis) [inline boot/kernel_os.go]: would use a *

Comment (bboozzoo) [reply boot/kernel_os.go]: This is done now.

File: boot/boot.go (hunk 4 above)
Comment (bboozzoo) [reply boot/boot.go]: Done as well.

File: boot/kernel_os.go (hunk 1 above)
Comment (anonymouse64) [reply boot/kernel_os.go]: ah I misunderstood your comment, I thought the issue was with passing `dev` to `boot.Kernel`, but you just didn't want `dev` in the struct

**Pull request merged**